src/core/theme/utils/defaults.ts

```diff
@@ -0,0 +1,42 @@
+import {
+  suomifiTheme,
+  SuomifiThemeProp,
+  defaultTokens,
+  TokensProp,
+  DefaultTokensProp,
+} from '../';
+
+const internalTokens = (tokens?: TokensProp) =>
+  !!tokens ? { ...defaultTokens, ...tokens } : defaultTokens;
+
+/**
+ * Check component props and do common defaulting
+ * - Check if tokens are given or use default ones
+ * - include internal tokens (THESE ARE NOT GEMERATED BY GIVEN TOKENS! but can be overridden by given tokens)
+ * @param props All component's props
+ */
+export const withSuomifiDefaults = <T extends { tokens: TokensProp }>({
```

Comment:
The chain up this point is super difficult to follow in my opinion. The use of tokens and theme with the component props in confusing since it's difficult to understand where those actually come from. If I understood correctly those are sometimes passed forward with passProps, which masks the names and actual props used later on.

The very least I would expect to see is a class diagram explaining the connections between theme, tokens, utils and components. Not sure if it would be possible to refactor and use a simpler approach to achieve the current functionality.

Reply:
Discussed elsewhere, but creating a diagram to dev readme. PassProps is understandable by Typescript and IDE that can say what it contains - would not write it open as it would make functionality a lot harder to fix and also typescript would need a lot complicated boilerplate syntax as it understands it now.